src/price_data.py

```diff
@@ -104,6 +106,60 @@ def _get_price_binance(self, base_asset: str, utc_time: datetime.datetime, quote
         average_price = total_cost / total_quantity
         return average_price
 
+    @misc.delayed
+    def _get_price_kraken(self, base_asset: str, utc_time: datetime.datetime, quote_asset: str, swapped_symbols: bool = False) -> float:
+        """Retrieve price from Kraken official REST API.
+
+        We select the data point closest to the desired timestamp (utc_time), but not newer than this timestamp.
+        For this we fetch one chunk of the trade history, starting one hour before this timestamp.
+        We then walk through the history until the closest timestamp match is found.
+        Otherwise, we start another 10 minutes earlier and try again, etc. …
+        (Exiting with a warning and zero price after hitting the arbitrarily chosen offset limit of 120 minutes.)
+
+        Documentation: https://www.kraken.com/features/api
+        """
+        target_time = int(utc_time.timestamp() * 1000)
+        root_url = "https://api.kraken.com/0/public/Trades"
+        pair = base_asset + quote_asset
+        pair = kraken_pair_map.get(pair, pair)
+
+        minutes_offset = 0
+        while minutes_offset < 120:
+            minutes_offset += 10
+
+            start_time, end_time = misc.get_offset_timestamps(utc_time, datetime.timedelta(minutes=minutes_offset))
+            since = start_time * 1000000
```

Comment:
> Why do we have to multiply?

The Kraken API is _really_ cumbersome. Timestamps are excepted in nanoseconds, thus the multiplication.

> And how is it possible to get from a timestamp to a trade id?

It is impossible. That why we need to scan for the closest timestamp before any particular order. I don't like it, and maybe I didn't understand it completely. But here's how I see it: We want to get the exchange rate of any particular asset at some particular time, without juggling the two matching Kraken account statement files at once.
And you seem to do a similar thing in the Binance case -- but you're averaging over some interval instead of finding the last exchange rate _before_ that particular trade.

Reply:
I asked about the trade id because the documentation says, that it takes a id for `since` instead of a timestamp. Or am I reading the documentation wrong?

> pair = asset pair to get trade data for
> since = return trade data since given id (optional.  exclusive)

https://www.kraken.com/features/api

Reply:
Hmm, I just tried it with a "txid" and an "ordertxid", both with and without specifying the underlying asset pair. But the parameter value is ignored! I guess the "ID" is really just the timestamp in nanoseconds. 

EDIT: Oh ok: `last = id to be used as since when polling for new trade data`. As this "last" value is clearly a timestamp, this leaves no doubt about the type of ID.

Reply:
I found an example [here](https://support.kraken.com/hc/en-us/articles/218198197-How-to-retrieve-historical-time-and-sales-trading-history-using-the-REST-API-Trades-endpoint-), which states that it is the timestamp in ns.

Reply:
Yep, that should settle it. Damn! Just inserting a trade ID would've been much nicer.